runtime/jvmti/jvmtiHook.c

```diff
@@ -539,6 +556,21 @@ jvmtiHookVMCheckpoint(J9HookInterface **hook, UDATA eventNum, void *eventData, v
 	TRACE_JVMTI_EVENT_RETURN(jvmtiHookVMCheckpoint);
 }
 
+static void
+jvmtiHookVMRestoreCRIUInit(J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData)
+{
+	J9RestoreEvent *data = eventData;
+	J9JVMTIEnv *j9env = userData;
+
+	Trc_JVMTI_jvmtiHookVMRestore_Entry();
+
+	data->currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(data->currentThread);
```

Comment:
This will be called while we are holding on to exclusive or safe point access. Im not sure this operation is safe.

D we need to release VM access

Reply:
`criuRestoreInitializeLib()` eventually releases VM access when it loads the agent library.
```
		vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
		vm->internalVMFunctions->acquireExclusiveVMAccess(currentThread);
...
		vm->internalVMFunctions->releaseExclusiveVMAccess(currentThread);
		vm->internalVMFunctions->internalExitVMToJNI(currentThread);
#0  allocateEnvironment (invocationJavaVM=0x7ffff0283af0, version=806682624, penv=0x7fffe0830aa0 <s>) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHelpers.cpp:382
#1  0x00007ffff56e4e70 in jvmtiHookGetEnv (hook=0x7ffff0012120, eventNum=18, eventData=0x7ffff7d35ac0, userData=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHook.c:1969
#2  0x00007ffff7f9c6ce in J9HookDispatch (hookInterface=0x7ffff0012120, taggedEventNum=18, eventData=0x7ffff7d35ac0) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/omr/util/hookable/hookable.cpp:237
#3  0x00007ffff79d5eca in GetEnv (jvm=0x7ffff0283af0, penv=0x7fffe0830aa0 <s>, version=806682624) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/vm/jniinv.c:898
#4  0x00007fffe07ec1f5 in Agent_OnLoad (vm=0x7ffff0283af0, options=0x7ffff0296b75 "transport=dt_socket,server=y,suspend=y", reserved=0x0) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:208
#5  0x00007ffff56f8565 in issueAgentOnLoadAttach (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98, options=0x7ffff0296b75 "transport=dt_socket,server=y,suspend=y", loadFunctionName=0x7ffff573255e "Agent_OnLoad", foundLoadFn=0x7ffff7d35c88)
    at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:519
#6  0x00007ffff56f8cac in loadAgentLibraryGeneric (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98, loadFunctionName=0x7ffff573255e "Agent_OnLoad", loadStatically=0, found=0x7ffff7d35c88, errorMessage=0x7ffff7d35c90)
    at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:632
#7  0x00007ffff56f9ae1 in loadAgentLibrary (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:862
#8  0x00007ffff56f79e9 in processAgentLibraryFromArgsList (vm=0x7ffff000fdc0, argsList=0x7ffff027d9f0, agentColon=0x7ffff573227b "-agentlib:", loadLibrary=1) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:190
#9  0x00007ffff56f7d2e in criuRestoreInitializeLib (vm=0x7ffff000fdc0, j9env=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:278
#10 0x00007ffff56dff07 in jvmtiHookVMRestoreCRIUInit (hook=0x7ffff0012120, eventNum=103, eventData=0x7ffff7d35fe0, userData=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHook.c:568
```

`criuRestoreInitializeLib()` is registered with `J9HOOK_VM_PREPARING_FOR_RESTORE` to ensure it is invoked before JIT hooks, and determine if a `JDWP` agent is specified in the restore option file.
`TRIGGER_J9HOOK_VM_PREPARING_FOR_RESTORE()` is fired before https://github.com/eclipse-openj9/openj9/blob/bfa3cfd037f9afb183486dd048bd9a9bf016d1e2/runtime/vm/CRIUHelpers.cpp#L1943

I am going to keep the `J9HOOK_VM_PREPARING_FOR_RESTORE` hook to process the restore option arguments and enable `J9VM_CRIU_IS_JDWP_ENABLED` accordingly, and move the actual agent loading after `releaseSafeOrExcusiveVMAccess()`.

Reply:
> to ensure it is invoked before JIT hooks

The JIT also registers against the `J9HOOK_VM_PREPARING_FOR_RESTORE` hook; is it a guarantee that the jit hook will be called after `criuRestoreInitializeLib`?

Reply:
Yes, with `J9HOOK_AGENTID_FIRST` at `J9HookRegisterWithCallSite(vmHook, J9HOOK_TAG_AGENT_ID | J9HOOK_VM_PREPARING_FOR_RESTORE, jvmtiHookVMRestoreCRIUInit, OMR_GET_CALLSITE(), jvmtiData, J9HOOK_AGENTID_FIRST)`.

Reply:
Hooked `J9HOOK_VM_PREPARING_FOR_RESTORE` event to determine if a `JDWP` agent is specified in the restore option file, also hooked `J9HOOK_VM_CRIU_RESTORE` to load the agent libraries which is triggered at https://github.com/eclipse-openj9/openj9/blob/4fbb9648b26ac6a2d5df22973d9b8feeeab580f3/runtime/vm/CRIUHelpers.cpp#L1943-L1944

@tajila this is ready for another look.